File: Example/Lua/MySampleTextPlugin.lua
--[[
enum class NameType
{
    None, Single, Multiple
};

struct Sentence {
    int index;
    std::string name;
    std::vector<std::string> names;
    std::string name_preview;
    std::vector<std::string> names_preview;
    std::string original_text;
    std::string pre_processed_text;
    std::string pre_translated_text;
    std::vector<std::string> problems;
    std::string translated_by;
    std::string translated_preview;
    std::string originalLinebreak;
    std::map<std::string, std::string> other_info;

    NameType nameType = NameType::None;
    Sentence* prev = nullptr;
    Sentence* next = nullptr;
    bool complete = false;
    bool notAnalyzeProblem = false;

    bool other_info_contains(const std::string& key);
    std::string other_info_get(const std::string& key);
    void other_info_set(const std::string& key, const std::string& value);
    std::tuple<std::vector<std::string>, std::vector<std::string>> other_info_get_all();
    void other_info_set_all(std::vector<std::string> keys, std::vector<std::string> values);
    void other_info_erase(const std::string& key);
    void other_info_clear();
    std::optional<std::string> problems_get_by_index(int index);
    bool problems_set_by_index(int index, const std::string& problem);
};

utils里的工具函数的签名详见 LuaMnager::registerCustomTypes 中绑定 utils 库的部分
]]

sourceLang_useTokenizer = false
targetLang_useTokenizer = true
--sourceLang_tokenizerBackend = "MeCab"
targetLang_tokenizerBackend = "pkuseg"
--sourceLang_meCabDictDir = "..."
--targetLang_spaCyModelName = "..."
--...

function checkConditionForRetranslKeysFunc(sentence)
    if sentence.index == 278 then 
        utils.logger:info("RetransFromLua")
    end
    return false
end

function checkConditionForSkipProblemsFunc(sentence)
    if sentence.index == 278 then 
        -- Example/Python/MySampleTextPlugin 中有对此函数的进一步介绍
        current_problem = ""
        for i, s in ipairs(sentence.problems) do
            if string.find(s, "^Current problem:") then 
                current_problem = string.sub(s, 17)
            else
                local success = sentence:problems_set_by_index(i-1, "My new problem")
                if success then
                    sentence:other_info_set("luaSuccess", "New problem suc")
                else
                    sentence:other_info_set("luaFail", "New problem fail")
                end
            end
        end
        sentence:other_info_set("luaCheck", "The 278th")
        sentence:other_info_set("luacp", current_problem) 
    end
    return false
end

function run(sentence)
    if sentence.index == 0 then
        sentence:other_info_set("MySampleKeyForFirstSentence", "MySampleValueForFirstSentence")
    end
    if sentence.next == nil and sentence.prev ~= nil and sentence.prev.nameType == NameType.None then
        sentence.problems:add("MySampleProblemForLastSentence")
        sentence.notAnalyzeProblem = true
    end
    if sentence.translated_preview == "久远紧锁眉头，脸上写着「骗人的吧，喂」。" then
        local wordpos_vec, entity_vec = utils.targetLang_tokenizeFunc(sentence.translated_preview)
        local tokens = utils.splitIntoTokens(wordpos_vec, sentence.translated_preview)
        local parts = {}
        for i, value in ipairs(wordpos_vec) do
            parts[i] = "[" .. value[1] .. "]"
        end
        local result = table.concat(parts, "#")
        sentence:other_info_set("tokens", result)
        utils.logger:info(string.format("测试 logger ，当前 index: %d", sentence.index))
    end
end

function init(projectDir)
    utils.logger:info(string.format("MySamplePlugin 初始化成功，projectDir: '%s'", projectDir))
end

function unload()
    local cmd = [[D:\GALGAME\linshi\天冥のコンキスタ\WORK\#toSjis.bat]]
    local cmd_acp = utils.ascii2Ascii(cmd, 65001, 0)
    local handle = io.popen(cmd_acp, 'r')
    if handle then
        local output = handle:read("*a")
        handle:close()
        utils.logger:info("--- BAT 脚本的输出内容 开始 ---\n" .. output .. "\n--- BAT 脚本的输出内容 结束 ---")
    else
        utils.logger:error("无法执行 BAT")
    end
end

File: Example/Python/MySampleTextPlugin.py
# Required: Import C++ binding module
import gpp_plugin_api as gpp
from pathlib import Path

# See Example/Lua/MySampleTextPlugin.lua for Sentence declaration details

# Global variables, injected by C++
# In Python, logger and tokenizeFunc are in the current module's globals, not in utils
logger = None

# sourceLang_useTokenizer = True
# sourceLang_tokenizerBackend = "MeCab"
# sourceLang_meCabDictDir = "..."
# sourceLang_spaCyModelName = "ja_core_news_sm"
# targetLang_spaCyModelName = "..."
# targetLang_stanzaLang = "..."
# sourceLang_tokenizeFunc = None
# ...

targetLang_useTokenizer = True
targetLang_tokenizerBackend = "pkuseg"
# If using provided tokenizer, must define tokenizeFunc first
targetLang_tokenizeFunc = None


def checkConditionForRetranslKeysFunc(se: gpp.Sentence) -> bool:
    if se.index == 278:
        logger.info("RetransFromPython")
    return False

def checkConditionForSkipProblemsFunc(se: gpp.Sentence) -> bool:
    if se.index == 278:
        # The se in retranslKey check is a copy, only for checking - modifications won't affect translation/output
        # But skipProblems check provides a reference to the actual se to be output
        # By inserting a custom flag problem via a custom post-processing plugin
        # and defining the first element in skipProblems as a full match check for this problem
        # You can even achieve running a plugin again after post-translation dict finishes
        # (difference: won't call init and unload)
        current_problem = ""
        for i, s in enumerate(se.problems):
            # After regex check passes, during Condition verification, the current problem has prefix 'Current problem:'
            # PS: DO NOT modify the problems object itself in any way within this check function - may cause C++ side exceptions
            # But you can modify element content indirectly via problems_set_by_index
            if s.startswith("Current problem:"):
                current_problem = s[16:]
            else:
                success = se.problems_set_by_index(i, "My new problem")
                if success:
                    se.other_info |= { f"pySuccess{i}": "New problem suc" }
                else:
                    se.other_info |= { f"pyFail{i}": "New problem fail" }

        se.other_info |= { "pythonCheck": "The 278th" }
        se.other_info |= { "pyCp": current_problem }
        logger.info("SkipProblemsFromPython")
    return False

def init(project_dir: Path):
    """
    Plugin initialization function, called once by C++.
    """
    logger.info(f"MySampleTextPluginFromLua initialized successfully, projectDir: {project_dir}")


def run(se: gpp.Sentence):
    """
    Main function to process each sentence.
    se is a proxy for a C++ Sentence object.
    """
    if se.translated_preview == "久远紧锁眉头，脸上写着「骗人的吧，喂」。":
        wordpos_vec, entity_vec = targetLang_tokenizeFunc(se.translated_preview)
        # Other utility functions (though not many) are still in utils
        tokens = gpp.utils.splitIntoTokens(wordpos_vec, se.translated_preview)
        parts = [f"[{value[0]}]" for value in wordpos_vec]
        result = "#".join(parts)
        # All containers are copies, direct insert/append etc. operations are ineffective
        se.other_info |= { "tokens_python": result }

def unload():
    pass
